packages/hardhat-core/src/internal/cli/ArgumentsParser.ts

```diff
@@ -102,6 +106,71 @@ export class ArgumentsParser {
         envVariableArguments,
         hardhatArguments
       ),
+      isCompileTask: taskName === TASK_COMPILE,
+      allUnparsedCLAs,
+    };
+  }
+
+  public parseScopeAndTaskNames(
+    allUnparsedCLAs: string[],
+    taskDefinitions: TasksMap,
+    scopeDefinitions: ScopesMap
+  ): {
+    scopeName?: string;
+    taskName: string;
+    unparsedCLAs: string[];
+  } {
+    let scopeName: string | undefined;
```

Comment:
I think this is good, but it could use a little bit of cleanup. For example you could put `const [firstCLA, secondCLA, ...restCLAs] = allUnparsedCLAs;` at the top to avoid repetition. Same with `const scopeDefinition = scopeDefinitions[firstCLA];` and maybe `taskDefinitions[firstCLA]`. Also what do you think of using early return here?